core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -509,10 +532,48 @@ public void kill() throws InterruptedException {
             }
             
             LOGGER.log(FINER, "Killing {0}", getPid());
+            killSoftly(getPid());
             p.kill();
             killByKiller();
         }
 
+        private void killSoftly(int pid) {
+            // send Ctrl+C to the process in the first iteration
+            // after that just wait for it to cease to exist
+            for (int i = 0; i < softkillWaitSeconds; ++i) {
```

Comment:
good comment form selckin on IRC: better to calculate the deadline = nowtimestamp+softkillWaitSeconds) and loop while (nowtimestamp < deadline)